### PR TITLE
CRD Status subresource support get+update+patch

### PR DESCRIPTION
**Pull request by roycaihw**
CRD Status previously only supports PUT and returns 405 on GET and PATCH

/assign @sttts 
/sig api-machinery

**Release note**:

```release-note
CustomResourceDefinitions Status subresource now supports GET and PATCH
```

### CODE REVIEW

**Comment by sttts** [comment]
We need a test for that in test/integration in the apiextensions-apiserver.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go [hunk 1]
```diff
@@ -160,12 +160,17 @@ type StatusREST struct {
 	store *genericregistry.Store
 }
 
-var _ = rest.Updater(&StatusREST{})
+var _ = rest.Patcher(&StatusREST{})
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go]: nit: keep the Updater assertion and add Patcher

Comment (roycaihw) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go]: Patcher assertion implies Updater assertion https://github.com/kubernetes/kubernetes/blob/c94efcea1d5f996b0d69b31472dff581b8a25be9/staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go#L252-L256

but we don't have a lot of Updater/Patcher assertion in our code base (grep result is 1 Updater assertion and 6 Patcher assertion), so I don't know what is good convention/style

**Comment by roycaihw** [comment]
Integration test added to check we expose the verbs

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go]: This is fine as is, then

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go [hunk 2]
```diff
@@ -799,3 +800,38 @@ func TestNameConflict(t *testing.T) {
...
+		Name(noxuDefinition.Name).
+		SubResource("status").
+		Do().
+		Into(result)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// make sure we don't get 405 Method Not Allowed from Patching CRD/status subresource
+	_, err = apiExtensionClient.ApiextensionsV1beta1().CustomResourceDefinitions().
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: nit: `if err = ...; err != nil`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go (hunk 1 above)
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go]: Can you add this line to the [custom resource `StatusREST`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go#L115) as well? It currently satisfies the `Patcher` interface implicitly and supports get/update/patch but I think explicitly mentioning it is nicer and safer.

Also, `StatusREST` for [deployments](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/apps/deployment/storage/storage.go#L117), [statefulsets](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/apps/statefulset/storage/storage.go#L98), etc does not implement `Patcher` explicitly as well...but I guess we should add it at least for the custom resource.

**Comment by lavalamp** [comment]
cc @mbohlool

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go (hunk 1 above)
Comment (roycaihw) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go]: added

**Comment by sttts** [comment]
@roycaihw  please squash. Then lgtm and approved.

**Pull request merged**